gpytorch/lazy/lazy_tensor.py

```diff
@@ -775,73 +775,58 @@ def cat_rows(self, cross_mat, new_mat, generate_roots=True, **root_decomp_kwargs
         # don't create one
-        does_not_have_roots = _is_in_cache_ignore_args(self, "root_decomposition") or _is_in_cache_ignore_args(
-            self, "root_inv_decomposition"
+        does_not_have_roots = any(
+            _is_in_cache_ignore_args(self, key) for key in ("root_inv_decomposition", "root_inv_decomposition")
         )
         if not generate_roots and not does_not_have_roots:
             return new_lazy_tensor
 
-        L = self.root_decomposition(**root_decomp_kwargs).root
-        l_is_triang = isinstance(L, TriangularLazyTensor)
-        L = L.evaluate()
-        L_inverse = self.root_inv_decomposition().root.evaluate()
-
-        m, n = L.shape[-2:]
-
-        if l_is_triang:
-            # The whole thing is triangular, we can just do two triangular solves
-            if isinstance(cross_mat, LazyTensor):
-                cross_mat = cross_mat.evaluate()
-            lower_left = torch.triangular_solve(cross_mat.transpose(-1, -2), L, upper=False).solution.transpose(-2, -1)
-        else:
-            lower_left = cross_mat.matmul(L_inverse)
-
-        schur = new_mat - lower_left.matmul(lower_left.transpose(-2, -1))
-        schur_root = lazify(schur).root_decomposition().root.evaluate()
-
-        # Form new root Z = [L 0; lower_left schur_root]
-        num_fant = schur_root.size(-2)
-        new_root = torch.zeros(*batch_shape, m + num_fant, n + num_fant, device=L.device, dtype=L.dtype)
-        new_root[..., :m, :n] = L
-        new_root[..., m:, : lower_left.shape[-1]] = lower_left
-        new_root[..., m:, n : (n + schur_root.shape[-1])] = schur_root
-
-        # Use pseudo-inverse of Z as new inv root
-        if l_is_triang:
-            # we know L is triangular, so inverting is a triangular solve agaist the identity
-            # we don't need the batch shape here, thanks to broadcasting
-            Eye = torch.eye(new_root.shape[-2], device=new_root.device, dtype=new_root.dtype)
-            new_inv_root = torch.triangular_solve(Eye, new_root, upper=False)[0]
-            new_inv_root = new_inv_root.transpose(-1, -2)
+        E = self.root_decomposition(**root_decomp_kwargs).root
+        m, n = E.shape[-2:]
+        R = self.root_inv_decomposition().root.evaluate()  # this is fast if L is triangular
+        F = B_ @ R
+        G = lazify(D - F.matmul(F.transpose(-2, -1))).root_decomposition().root.evaluate()
+
+        # Form new root Z = [E 0; F G]
+        num_fant = G.size(-2)
+        new_root = torch.zeros(*batch_shape, m + num_fant, n + num_fant, device=E.device, dtype=E.dtype)
+        new_root[..., :m, :n] = E.evaluate()
+        new_root[..., m:, : F.shape[-1]] = F
+        new_root[..., m:, n : (n + G.shape[-1])] = G
+
+        if isinstance(E, TriangularLazyTensor) and isinstance(G, TriangularLazyTensor):
+            # make sure these are actually upper triangular
+            if getattr(E, "upper", False) or getattr(G, "upper", False):
+                raise NotImplementedError
```

Comment:
Can this just be resolved with a transpose? Perhaps something like `new_root = TriangularLazyTensor(new_root, upper=False)`?

Reply:
Well so we assume from the start that we're going to build a block lower triangular matrix. Here we're in the case where `E` or `G` are triangular, but if either of them is upper triangular then there is no way the whole matrix is going to be. Does this make sense? I don't think we currently would ever get into that situation but I figured better safe than sorry here.

Reply:
Ahh yeah, good point. The proper thing to do is probably to use the transpose before we even get to this point then, but I think it's probably unlikely to hit it.

Reply:
yeah, we can fix it when we run into it, at least it's breaking loudly